node/src/types/sync_leap.rs

```diff
@@ -290,20 +295,1676 @@ impl FetchItem for SyncLeap {
+
+        // Assert only if correct eras are selected, since the the
+        // `should_return_era_validator_weights_for_correct_sync_leap` test already covers the
+        // actual weight validation.
+
+        let actual_eras: BTreeSet<u64> = sync_leap
+            .era_validator_weights(fault_tolerance_fraction)
+            .map(|era_validator_weights| era_validator_weights.era_id().into())
+            .collect();
+        let mut expected_eras: BTreeSet<u64> = BTreeSet::new();
+        // Expect successors of the eras of switch blocks.
+        expected_eras.extend([2, 3, 4]);
+        assert_eq!(expected_eras, actual_eras);
+    }
+
+    #[test]
+    fn era_validator_weights_without_genesis_with_switch_block_preceding_immediate_switch_block() {
+        let mut rng = TestRng::new();
+
+        let trusted_block = Block::random_non_switch_block(&mut rng);
+
+        let (
+            signed_block_header_with_metadata_1,
+            signed_block_header_with_metadata_2,
+            signed_block_header_with_metadata_3,
+        ) = make_three_switch_blocks_at_era_and_height(rng, (1, 10), (2, 20), (3, 21));
+
+        let sync_leap = SyncLeap {
+            trusted_ancestor_only: false,
+            trusted_block_header: trusted_block.header().clone(),
+            trusted_ancestor_headers: vec![],
+            signed_block_headers: vec![
+                signed_block_header_with_metadata_1,
+                signed_block_header_with_metadata_2,
+                signed_block_header_with_metadata_3,
+            ],
+        };
+
+        let fault_tolerance_fraction = Ratio::new_raw(1, 3);
+
+        // Assert only if correct eras are selected, since the the
+        // `should_return_era_validator_weights_for_correct_sync_leap` test already covers the
+        // actual weight validation.
+
+        let actual_eras: BTreeSet<u64> = sync_leap
+            .era_validator_weights(fault_tolerance_fraction)
+            .map(|era_validator_weights| era_validator_weights.era_id().into())
+            .collect();
+        let mut expected_eras: BTreeSet<u64> = BTreeSet::new();
+
+        // Block #1 (era=1, height=10)
+        // Block #2 (era=2, height=20) - block preceding immediate switch block
+        // Block #3 (era=3, height=21) - immediate switch block.
+        // Expect the successor of block #2 to be not present.
+        expected_eras.extend([2, 4]);
+        assert_eq!(expected_eras, actual_eras);
+    }
+
+    #[test]
+    fn era_validator_weights_with_genesis_without_switch_block_preceding_immediate_switch_block() {
+        let mut rng = TestRng::new();
+
+        let trusted_block = Block::random_non_switch_block(&mut rng);
+
+        let (
+            signed_block_header_with_metadata_1,
+            signed_block_header_with_metadata_2,
+            signed_block_header_with_metadata_3,
+        ) = make_three_switch_blocks_at_era_and_height(rng, (0, 0), (2, 20), (3, 30));
```

Comment:
I guess I'd have used `(0, 0), (1, 10), (2, 20)` to make it slightly more realistic, but not a big deal :)